src/Graph/DOT.elm

```diff
@@ -0,0 +1,172 @@
+-}
+type Rankdir
+    = TB
+    | LR
+    | BT
+    | RL
+
+
+{-| A blank `Styles` record to build from to define your own styles.
+
+    myStyles =
+        { defaultStyles
+            | node = "shape=box, color=blue, style=\"rounded, filled\""
+        }
+
+-}
+defaultStyles : Styles
+defaultStyles =
+    Styles TB "" "" ""
+
+
+{-| Same as `output`, but allows you to add attrs to the graph.
+These attrs will be applied to the entire graph.
+-}
+outputWithStyles : Styles -> (n -> Maybe String) -> (e -> Maybe String) -> Graph n e -> String
+outputWithStyles styles mapNode mapEdge graph =
+    let
+        labelOnly maybeLabel =
+            case maybeLabel of
+                Nothing ->
+                    Dict.empty
+
+                Just l ->
+                    Dict.singleton "label" l
+    in
+    outputWithStylesAndAttributes styles (labelOnly << mapNode) (labelOnly << mapEdge) graph
+
+
+{-| Same as `outputWithStyles`, but allows each node and edge to include its
+own attrs. Note that you must supply a conversion function for node and edge
+labels that return a `Dict String String` of the attribute mappings.
+
+Note that you have to take care of setting the appropriate node and edge labels
+yourself.
+
+-}
+outputWithStylesAndAttributes :
+    Styles
+    -> (n -> Dict String String)
+    -> (e -> Dict String String)
+    -> Graph n e
+    -> String
+outputWithStylesAndAttributes styles nodeAttrs edgeAttrs graph =
+    let
+        attrAssocs : Dict String String -> String
+        attrAssocs =
+            Dict.toList
+                >> List.map (\( k, v ) -> k ++ "=\"" ++ v ++ "\"")
+                >> String.join ", "
+
+        makeAttrs : Dict String String -> String
+        makeAttrs d =
+            if Dict.isEmpty d then
+                ""
+            else
+                " [" ++ attrAssocs d ++ "]"
+
+        edges =
+            let
+                compareEdge a b =
+                    case compare a.from b.from of
+                        LT ->
+                            LT
+
+                        GT ->
+                            GT
+
+                        EQ ->
+                            compare a.to b.to
+            in
+            Graph.edges graph
+                |> List.sortWith compareEdge
+
+        nodes =
+            Graph.nodes graph
+
+        edgesString =
+            List.map edge edges
+                |> String.join ";\n"
+
+        edge e =
+            "  "
+                ++ Basics.toString e.from
+                ++ " -> "
+                ++ Basics.toString e.to
+                ++ makeAttrs (edgeAttrs e.label)
+
+        nodesString =
+            List.map node nodes
+                |> String.join ";\n"
+
+        node n =
+            "  "
+                ++ Basics.toString n.id
+                ++ makeAttrs (nodeAttrs n.label)
+    in
+    "digraph G {\n"
+        ++ ("  rankdir=" ++ toString styles.rankdir ++ ";\n")
+        ++ ("  graph [" ++ styles.graph ++ "]" ++ ";\n")
```

Comment:
Maybe get rid of the ++ between `"]" ++ ";\n"` to make things tiny bit more efficient.